recirq/quantum_chess/circuit_transformer.py

```diff
@@ -305,6 +307,391 @@ def transform(self, circuit: cirq.Circuit) -> cirq.Circuit:
+        """Returns an adjacency graph of logical qubits of the circuit.
+
+        Uses the heuristic of adding an edge between the nodes of each
+        disjoint component that are least connected if the graph contains more
+        than one connected component.
+
+        Each edge is a tuple containing an adjacent node and the index of the
+        moment at which the operation occurs.
+
+        Arg:
+          circuit: The circuit from which to build a logical qubits graph.
+        """
+        g = defaultdict(list)
+
+        # Build an adjacency graph based on the circuit.
+        for i, m in enumerate(circuit):
+            for op in m:
+                if len(op.qubits) == 2:
+                    q1, q2 = op.qubits
+                    q1_neighbors = [n[0] for n in g[q1]]
+                    if q2 not in q1_neighbors:
+                        g[q1].append((q2, i))
+                    q2_neighbors = [n[0] for n in g[q2]]
+                    if q1 not in q2_neighbors:
+                        g[q2].append((q1, i))
+
+        # Find the connected components in the graph.
+        components = deque()
+        visited = set()
+        for q in g:
+            if q not in visited:
+                components.append(self.traverse(g, q, visited))
+
+        if len(components) == 1:
+            return g
+
+        # Connect disjoint components by adding an edge between the nodes of
+        # each disjoint component that are least connected.
+        while len(components) > 1:
+            first_comp = components.pop()
+            first_q = self.get_least_connected_qubit(g, first_comp)
+            second_comp = components.pop()
+            second_q = self.get_least_connected_qubit(g, second_comp)
+
+            # Add an edge between the two least connected nodes, and set its
+            # moment index to infinity since this edge doesn't represent a
+            # actual gate in the circuit.
+            g[first_q].append((second_q, math.inf))
+            g[second_q].append((first_q, math.inf))
+
+            # Combine the two components and add it back to the components
+            # deque to continue connecting disjoint components.
+            first_comp += second_comp
+            components.append(first_comp)
+
+        return g
+
+    def find_graph_center(
+        self,
+        g: Union[
+            Dict[cirq.GridQubit, List[cirq.GridQubit]],
+            Dict[cirq.Qid, List[Tuple[cirq.Qid, int]]],
+        ],
+    ) -> Union[cirq.GridQubit, cirq.Qid]:
+        """Returns a qubit that is a graph center.
+
+        Uses the Floyd-Warshall algorithm to calculate the length of the
+        shortest path between each pair of nodes. Then, finds the graph center
+        such that the length of the shortest path to the farthest node is the
+        smallest. Returns the first graph center if there are multiple.
+
+        Args:
+          g: A physical qubits graph or a logical qubits graph.
+        """
```

Comment:
should we move the floyd-warshall implementation into a separate file? That may be better organizationally. And since it is a fairly generic algorithm, it seems like something that we could add test coverage for independently of the rest of the initial mapping implementation

Reply:
i agree the floyd-warshall algorithm and the various bfs algorithms are fairly generic, but their implementation all have to be aware of whether it's graph param is a logical qubit graph (where edges contain moment info) or physical qubit graph or either, so i think it might make more sense to keep them in the same class.

but yah, it's def worth adding test coverage independently for these more generic algorithms; maybe we can keep the test files separate? kind of like what we have right now in dynamic_look_ahead_heuristic_circuit_transformer_test.py, but with more / better test cases?

Reply:
Agreed, lets add smaller focused files if we can. Besides the tests I'm also a bit concerned about the potential for circuit_transformer.py to explode in size if we add more transformers, but should be fine for now.

Reply:
by the way, I realized another reason it would be nice to refactor out the floyd-warshall implementation: the swap update step needs the precomputed node-to-node shortest path distances, too. We can leave that for a future pull request though.

Reply:
sg, i refactored it out to a separate function!